quickwit/quickwit-search/src/leaf_cache.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use quickwit_proto::{LeafSearchResponse, SearchRequest, SplitIdAndFooterOffsets};
+use quickwit_storage::MemorySizedCache;
+
+pub struct LeafSearchCache {
+    content: MemorySizedCache<CacheKey, LeafSearchResponse>,
+}
+
+impl LeafSearchCache {
+    pub fn new(capacity: usize) -> LeafSearchCache {
+        LeafSearchCache {
+            content: MemorySizedCache::with_capacity_in_bytes(
+                capacity,
+                &quickwit_storage::STORAGE_METRICS.partial_request_cache,
+            ),
+        }
+    }
+    pub fn get(
+        &self,
+        split_info: SplitIdAndFooterOffsets,
+        search_request: SearchRequest,
+    ) -> Option<LeafSearchResponse> {
+        let key = CacheKey::from_split_meta_and_request(split_info, search_request);
+        self.content.get(&key)
+    }
+
+    pub fn put(
+        &self,
+        split_info: SplitIdAndFooterOffsets,
+        search_request: SearchRequest,
+        result: LeafSearchResponse,
+    ) {
+        let key = CacheKey::from_split_meta_and_request(split_info, search_request);
+
+        self.content.put(key, result);
+    }
+}
+
+#[derive(Debug, Hash, PartialEq, Eq)]
+struct CacheKey {
+    split_id: String,
+    request: SearchRequest,
+    request_time_range: Range,
```

Comment:
isn't the request_time_range part of SearchRequest?

Reply:
ok so my understanding is that it has been cropped to match the split, to increase cache hits.

This should be explained in a comment.